config/cmake/ConfigureChecks.cmake

```diff
@@ -872,93 +872,113 @@ H5ConversionTests (${HDF_PREFIX}_LLONG_TO_LDOUBLE_CORRECT TRUE "Checking IF corr
-  CHECK_SYMBOL_EXISTS (FLT16_MAX_10_EXP "float.h" h5_have_flt16_max_10_exp)
-  CHECK_SYMBOL_EXISTS (FLT16_MANT_DIG "float.h" h5_have_flt16_mant_dig)
-
-  if (h5_have_flt16_epsilon AND h5_have_flt16_min AND
-      h5_have_flt16_max AND h5_have_flt16_min_10_exp AND
-      h5_have_flt16_max_10_exp AND h5_have_flt16_mant_dig)
-    # Some compilers like OneAPI on Windows appear to detect _Float16 support
-    # properly up to this point, and, in the absence of any architecture-specific
-    # tuning compiler flags, will generate code for H5Tconv.c that performs
-    # software conversions on _Float16 variables with compiler-internal functions
-    # such as __extendhfsf2, __truncsfhf2, or __truncdfhf2. However, these
-    # compilers will fail to link these functions into the build for currently
-    # unknown reasons and cause the build to fail. Since these are compiler-internal
-    # functions that we don't appear to have much control over, let's try to
-    # compile a program that will generate these functions to check for _Float16
-    # support. If we fail to compile this program, we will simply disable
-    # _Float16 support for the time being.
-
-    # Some compilers, notably AppleClang on MacOS 12, will succeed in the
-    # configure check below when optimization flags like -O3 are manually
-    # passed in CMAKE_C_FLAGS. However, the build will then fail when it
-    # reaches compilation of H5Tconv.c because of the issue mentioned above.
-    # MacOS 13 appears to have fixed this, but, just to be sure, backup and
-    # clear CMAKE_C_FLAGS before performing these configure checks.
-    set (cmake_c_flags_backup "${CMAKE_C_FLAGS}")
-    set (CMAKE_C_FLAGS "")
-
-    H5ConversionTests (
-        ${HDF_PREFIX}_FLOAT16_CONVERSION_FUNCS_LINK
-        FALSE
-        "Checking if compiler can convert _Float16 type with casts"
-    )
-
-    set (CMAKE_C_FLAGS "${cmake_c_flags_backup}")
-
-    if (${${HDF_PREFIX}_FLOAT16_CONVERSION_FUNCS_LINK})
-      # Finally, MacOS 13 appears to have a bug specifically when converting
-      # long double values to _Float16. Release builds of the dt_arith test
-      # would cause any assignments to a _Float16 variable to be elided,
-      # whereas Debug builds would perform incorrect hardware conversions by
-      # simply chopping off all the bytes of the value except for the first 2.
-      # These tests pass on MacOS 14, so let's perform a quick test to check
-      # if the hardware conversion is done correctly.
+# Option to enable or disable all non-standard features. Specific features can
+# be enabled or disabled with their respective options below
+option (HDF5_ENABLE_NONSTANDARD_FEATURES "Enable support for non-standard programming language features" ON)
+# Options for enabling or disabling individual features
+option (HDF5_ENABLE_NONSTANDARD_FEATURE_FLOAT16 "Enable support for _Float16 C datatype" ${HDF5_ENABLE_NONSTANDARD_FEATURES})
```

Comment:
@byrnHDF I couldn't figure out a nice way to make it so that turning off `HDF5_ENABLE_NONSTANDARD_FEATURES` in a GUI will also turn off `HDF5_ENABLE_NONSTANDARD_FEATURE_FLOAT16` initially and vice versa, but for most cases this tends to just work.

Reply:
I think this is a really good solution - can still be changed by command line.